### DISCUSSION TITLE
How to set the reuse factor

### DISCUSSION THREAD

**Discussion by davidecarini** [fastmachinelearning/hls4ml#775] [General]
Hello to everyone!

If i have this CNN, there is a way to find the right values to set the reuse factor for every layer ?
For example, i can use 1 as reuse factor for the first Conv2D layer but what is the value of the reuse factor of the second layer? 

 Layer (type)                Output Shape              Param #   
=================================================================
 convolution_layer_1 (Conv2D  (None, 26, 26, 32)       320       
 )                                                               
                                                                 
 ReLu_1 (Activation)         (None, 26, 26, 32)        0         
                                                                 
 max_pooling_layer_1 (MaxPoo  (None, 13, 13, 32)       0         
 ling2D)                                                         
                                                                 
 convolution_layer_2 (Conv2D  (None, 11, 11, 14)       4046      
 )                                                               
                                                                 
 ReLu_2 (Activation)         (None, 11, 11, 14)        0         
                                                                 
 max_pooling_layer_2 (MaxPoo  (None, 5, 5, 14)         0         
 ling2D)                                                         
                                                                 
 flatten_layer (Flatten)     (None, 350)               0         
                                                                 
 dropout_layer (Dropout)     (None, 350)               0         
                                                                 
 dense_layer (Dense)         (None, 10)                3510      
                                                                 
 softmax (Activation)        (None, 10)                0         
                                                                 
=================================================================
Total params: 7,876
Trainable params: 7,876
Non-trainable params: 0

**Comment by bo3z** [comment]
In general, the reuse factor (RF) for every layer will depend on preferences, such as latency vs. resources utilised and target architecture. A RF equal to one will mean all multiplications are executed in parallel. So, in general, for a layer of _n_mult_ multiplications and reuse factor of _reuse_factor_, we expect the layer to use `n_mult / reuse_factor` DSPs. This number can vary sometimes, due to various reasons - e.g. for low-precision (less than 9 bits), the HLS compiler will implement multiplication through combinatorial logic (LUT), rather than DSPs. The number of expected DSPs can also vary, depending on the DSP operating mode, target device etc. But, in general, this estimate is very accurate. Finally, there are some savings when your weight matrix is sparse and RF = 1 (however, when RF > 1, this is not the case)

For Dense layers, there is `n_in * n_out` multiplications, thus calculating the entire output after all the multiplications are done. For Conv2D layers there is `n_filt * n_chan * filt_width * filt_height` mutliplications for one output pixel (across all the output channels). By default, one output pixel is computed, before proceeding to the next one, but this could be tuned. 

In general, increasing RF will mean splitting the multiplications across more clock cycles - RF = 1 is equivalent to all multiplications done in parallel and RF = 4 would mean the multiplications are done across four clock cycles, with (approx.) four times less DSPs (as well as a reduction in other resources, but this reduction is not linear). So for a Dense layer, the expected latency is equal to the RF (plus some overhead to retrieve the weights from BRAM, if using Resource strategy). For Conv2D layers, the expected latency is proportional to the product of output pixels (height x width) and the RF.

Depending on your target device, latency constraints etc. the RF should be set accordingly for every layer. In your case, the model is fairly small and very high-end FPGAs (~8k DSPs) could accomodate the entire model with RF = 1. However, if the goal is to save resources - it might be worth increasing RF.

Examples on changing reuse factor, precision etc, as well as a more in-depth explanation on RF and DSPs used, are available in the [tutorial](https://github.com/fastmachinelearning/hls4ml-tutorial/blob/main/part2_advanced_config.ipynb).

One possible approach is to generate the configuration (Python dict) from the model, given a default reuse factor and then overwrite it for individual layers:
```
default_reuse_factor = 2
hls_config = config_from_keras_model(baseline_model, default_reuse_factor=default_reuse_factor)
hls_config['LayerName']['convolution_layer_2']['ReuseFactor'] = 4
```

The above code will generate an hls4ml config (a Python dictionary with all the necessary information for model conversion) with a default reuse factor of 2 for every layer. Then, the reuse factor can be overwritten, by changing the config for specific layers in the dictionary - `hls_config['LayerName'][layer_name]['ReuseFactor'] = new_reuse_factor`. In the above example, default reuse factor for all layers is 2 and in the second Conv2D layer it is 4. If not specified, the default reuse factor is set to 1 (for all layers)

**Reply by davidecarini** [reply]
Thank you for the answer! My problem is that if I set the RF = 1 the synthesis blocks during the scheduling of the second Conv2D layer and I don't understand the why. To solve the problem i had to set the RF of the second Conv2D to 288. There is an explanation to that behavior?  The target FPGA is the Alveo U280

**Reply by bo3z** [reply]
Without any specific logs of the synthesis it is hard to say - does synthesis simply block or does it fail? If it blocks, it might just be a matter of waiting longer - depending on your compiler, target architecture, model etc. synthesis can take very long. Also what version of hls4ml are you using? Is it installed from pip or from the main GitHub branch? The main branch is recommended as it contains many updates, over the pip version.

In general, Conv2D layers are very computationally intensive - it might be worth tweaking parameters, such as `io_type` (`io_stream` is recommended for CNNs, per [tutorial](https://github.com/fastmachinelearning/hls4ml-tutorial/blob/main/part6_cnns.ipynb), but `io_parallel` might work for smaller CNNs) or the strategy - `Resource` or `Latency`. If you check the tutorial on CNNs - a somewhat larger CNN should be synthesizeable with a low RF and `io_stream`

**Reply by davidecarini** [reply]
I use the version 0.6.0. The synthesis doesn't fail but it blocks for a long time that the terminal goes in broken pipe (because i work on a server using VPN). 

This is the config of hls4ml: 
hls4ml.model.optimizer.get_optimizer('output_rounding_saturation_mode').configure(layers=['Activation'], rounding_mode='AP_RND', saturation_mode='AP_SAT')
    Trace = True
    '''
    plots=hls4ml.model.profiling.numerical(model=model, hls_model=hls_model, X=testX[:1000]) #-> Non funziona (dynamic library problem)
    for i,plot in enumerate(plots):
        plot.savefig(f'Plots/hls4mlPlots{i}.png')
    '''
    config = hls4ml.utils.config_from_keras_model(model, granularity='name')
    config['Model'] = {}
    config['Model']['ReuseFactor'] = 1
    config['Model']['Strategy'] = 'Resource'
    config['Model']['Precision'] = 'ap_fixed<16,6>'
    
    #Reuse factor per layer
   
    config['LayerName']['convolution_layer_1']['ReuseFactor'] = 1
    config['LayerName']['convolution_layer_2']['ReuseFactor'] = 288
   
    config['LayerName']['dense_layer']['ReuseFactor'] = 1

    config['LayerName']['softmax']['Strategy']    = 'Stable'


    for layer in config['LayerName'].keys():
        config['LayerName'][layer]['Trace'] = True


    cfg = hls4ml.converters.create_config(board='alveo-u280', clock_period= 5, part='xcu280-fsvh2892-2L-e', backend='VivadoAccelerator')
    cfg['HLSConfig'] = config

    cfg['AcceleratorConfig']['Driver']    = 'python'
    cfg['AcceleratorConfig']['Board']     = 'alveo-u280'
    cfg['AcceleratorConfig']['Interface'] = 'axi_stream'
    cfg['AcceleratorConfig']['Precision']['Input']  = 'float'
    cfg['AcceleratorConfig']['Precision']['Output'] = 'float'

    cfg['IOType']= 'io_stream'
    cfg['KerasModel'] = model
    cfg['OutputDir'] = PROJECT_PATH + 'CNN/hls4ml_prj'
   
    
    hls_model = hls4ml.converters.keras_to_hls(cfg)
    hls4ml.model.optimizer.get_optimizer('output_rounding_saturation_mode').configure(layers=[])
    #hls4ml.model.profiling.numerical(model=model, hls_model=hls_model, X=testX[:1000])



The vivado log is attached. 
[vivado_log.pdf.odt](https://github.com/fastmachinelearning/hls4ml/files/11332422/vivado_log.pdf.odt)

**Reply by bo3z** [reply]
I haven't worked a lot with the accelerator backend, so someone else from the community might have a better answer for this specific example.

I would recommend two things - installing hls4ml from the main git branch, rather than pip. While I think streaming CNNs haven't changed significantly since the last release, it might solve the bug. 

Secondly, when working on servers, there are ways to make a process a continue after disconnecting from the server (e.g. due to internet going down or broken pipe). This will let you continue long synthesis without interruption. One common way is using the `screen` command in Linux (there are other commands as well). Using these commands will depend on your server setup and rules, so I recommend checking what is allowed/installed on your specific server. Also, if a process gets stuck, you might have to interrupt it manually (as it continues even after you disconnected).

As for the specific example - there is a limit on loop unrolling with Vivado, equal to 4,096. So in general, if there are more than 4,096 multiplications in a single layer - synthesis should fail. Other than that, have you tried a different RF for the second layer - e.g. 8 or 16? Why 288?

**Reply by davidecarini** [reply]
I try until 144 but the problem is the same. 288 is the first RF value for which the scheduling proceeds without problem and in a little time. The first layer has 320 parameters while the second has 4046 parameters. I tried others CNNs. If the #parameters of a single layer is > 4096, there is an error during the unrolling but the cause is what you are explained in your answer. I don't understand the why of this stuck situation when all the layers have #parameters < 4096. With RF = 288, the scheduling goes until the end and the process is very fast. If i decrease the RF to 144 the synthesis stuck in reuse loop also if I wait more than 2 hours. I think that also i wait other time, the problem remains.

**Reply by bo3z** [reply]
It could simply be the server time-out - the lower the RF, the more scheduling the compiler needs to do. Again, depending on your server resources, target device etc. HLS synthesis can take very long, even several hours (3-4) is relatively common. It might be worth exploring ways of keeping the server connection running.

**Reply by davidecarini** [reply]
Ok, thus your answer is that the synthesis could work also with RF = 1 for all the layers and the problem is the server connection time out?

**Reply by bo3z** [reply]
Not sure if RF = 1 will fit on the board, but it is worth trying a longer time-out / running the process after disconnecting from the server. Given larger models (e.g. in the tutorial) succeeded synthesis with lower RF, it should certainly be possible. Alternatively, depending on your dataset, maybe reduce model complexity by reducing the number of filters in the first layer?

**Reply by coldCecemel** [reply]
Why is that your available reusefactors change depending on the chosen strategy? (resource or latency)

**Reply by bo3z** [reply]
ReuseFactor doesn't depend on the strategy, only on layer input and output shapes.
